pkg/resource/v1/certconfig/desired.go

```diff
@@ -2,8 +2,347 @@ package certconfig
 
 import (
 	"context"
+	"net/url"
+	"strings"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/certs"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/versionbundle"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	"github.com/giantswarm/cluster-operator/pkg/label"
+	"github.com/giantswarm/cluster-operator/pkg/resource/v1/certconfig/key"
+)
+
+const (
+	certAPIVersion = "core.giantswarm.io"
+	certKind       = "CertConfig"
+	certOperatorID = "cert-operator"
+
+	loopbackIP = "127.0.0.1"
+
+	// systemMastersOrganization is the RBAC kubernetes admin group.
+	systemMastersOrganization = "system:masters"
+)
+
+var (
+	kubeAltNames = []string{
+		"kubernetes",
+		"kubernetes.default",
+		"kubernetes.default.svc",
+		"kubernetes.default.svc.cluster.local",
+	}
 )
 
+// GetDesiredState returns all desired CertConfigs for managed certificates.
 func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
-	return nil, nil
+	clusterGuestConfig, err := r.toClusterGuestConfigFunc(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	clusterConfig, err := prepareClusterConfig(*r.baseClusterConfig, *clusterGuestConfig)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredCertConfigs := make([]*v1alpha1.CertConfig, 0, len(managedCertificates))
+	for _, mc := range managedCertificates {
+		certConfig := mc.certConfigFactoryFunc(clusterConfig, mc.name, r.projectName)
+		desiredCertConfigs = append(desiredCertConfigs, certConfig)
+	}
+
+	return desiredCertConfigs, nil
+}
+
+func prepareClusterConfig(baseClusterConfig cluster.Config, clusterGuestConfig v1alpha1.ClusterGuestConfig) (*cluster.Config, error) {
+	var err error
+
+	// Copy baseClusterConfig as basis and supplement it with information from
+	// clusterGuestConfig.
+	clusterConfig := baseClusterConfig
+
+	clusterConfig.ClusterID = key.ClusterID(clusterGuestConfig)
+
+	clusterConfig.Domain.API, err = newServerDomain(key.APIEndpoint(clusterGuestConfig), string(certs.APICert))
```

Comment:
This chain of domain construction looks a bit dreadful, but couldn't quickly figure out better way to pass information without creating terrible function signature to `certConfigFactoryFunc`.